spec/unit/cloudkeeper/aws/cloud_spec.rb

```diff
@@ -0,0 +1,161 @@
+require 'spec_helper'
+
+module Cloudkeeper
+  module Aws
+    describe Cloud do
+      let(:s3) do
+        ::Aws::S3::Resource.new(stub_responses: true)
+      end
+
+      let(:ec2) do
+        ::Aws::EC2::Client.new(stub_responses: true)
+      end
+
+      let(:cloud) do
+        described_class.new(s3service: s3, ec2service: ec2)
+      end
+
+      describe '.upload_data' do
+        let(:filename) { 'test' }
+        let(:data) { 'Dusan is cool' }
+
+        context 'with empty bucket' do
+          before do
+            s3.client.stub_responses(:head_object,
+                                     status_code: 404, headers: {}, body: '')
+          end
+
+          it 'uploads data' do
+            expect do
+              cloud.upload_data(filename) \
+                { |write_stream| write_stream << data }
+            end.not_to raise_error
+          end
+        end
+
+        context 'with existing file' do
+          it 'raises error' do
+            expect do
+              cloud.upload_data(filename) \
+                { |write_stream| write_stream << data }
+            end.to raise_error(Cloudkeeper::Aws::Errors::BackendError)
+          end
+        end
+      end
+
+      describe '.poll_import_task' do
+        let(:import_id) { 'ii-69' }
+
+        context 'with status \'deleted\'' do
+          before do
+            ec2.stub_responses(:describe_import_image_tasks,
+                               import_image_tasks: [{ status: 'deleted' }])
+          end
+
+          it 'returns false' do
+            expect(cloud.poll_import_task(import_id)).to be(false)
+          end
+        end
+
+        context 'with status \'completed\'' do
+          before do
+            ec2.stub_responses(:describe_import_image_tasks,
+                               import_image_tasks: [{ status: 'completed' }])
+          end
+
+          it 'returns true' do
+            expect(cloud.poll_import_task(import_id)).to be(true)
+          end
+        end
+      end
+
+      describe '.search_tags' do
+        context 'with no resources' do
+          before do
+            stub_tags = []
+            ec2.stub_responses(:describe_tags, lambda do |context|
```

Comment:
Unless it's a new superfast low cost algorithm to solve world's hunger I shouldn't need a PhD. to understand it.

Reply:
I'll try...